service/history/api/multioperation/api.go

```diff
@@ -0,0 +1,201 @@
+
+	// grab current Workflow lease, if there is one
+	currentWorkflowLease, err := workflowConsistencyChecker.GetWorkflowLease(
+		ctx,
+		nil,
+		api.BypassMutableStateConsistencyPredicate,
+		definition.NewWorkflowKey(
+			string(namespaceID),
+			startReq.StartRequest.WorkflowId,
+			"",
+		),
+		workflow.LockPriorityHigh,
+	)
+	switch err.(type) {
+	case nil:
+		defer func() { currentWorkflowLease.GetReleaseFn()(retError) }()
+	case *serviceerror.NotFound:
+		currentWorkflowLease = nil
+	default:
+		return nil, err
+	}
+
+	if currentWorkflowLease != nil {
+		return nil, serviceerror.NewUnimplemented("not implemented yet")
+	}
+
+	starter, err := startworkflow.NewStarter(
+		shardContext,
+		workflowConsistencyChecker,
+		tokenSerializer,
+		visibilityManager,
+		startReq,
+	)
+	if err != nil {
+		return makeResponse(
+			&historyservicepb.WorkflowOperationResult{
+				Result: &historyservicepb.WorkflowOperationResult_Error{
+					// TODO: improve error reporting
+					Error: &status.Status{Code: code.Code_value[code.Code_INTERNAL.String()]},
+				},
+			},
+			&historyservicepb.WorkflowOperationResult{
+				Result: &historyservicepb.WorkflowOperationResult_Error{
+					Error: &status.Status{Code: code.Code_value[code.Code_ABORTED.String()]},
+				},
+			}), nil
+	}
+
+	var workflowCtx workflow.Context
+	var updateErr error
+	updater := updateworkflow.NewUpdater(
+		shardContext,
+		workflowConsistencyChecker,
+		matchingClient,
+		updateReq,
+	)
+	startResp, err := starter.Invoke(
+		ctx,
+		func(lease api.WorkflowLease) error {
+			workflowCtx = lease.GetContext()
+			_, updateErr = updater.Apply(ctx, lease)
+			return updateErr
+		},
+	)
+	if err != nil {
+		return makeResponse(
+			&historyservicepb.WorkflowOperationResult{
+				Result: &historyservicepb.WorkflowOperationResult_Error{
+					Error: &status.Status{Code: code.Code_value[code.Code_ABORTED.String()]},
+				},
+			},
+			&historyservicepb.WorkflowOperationResult{
+				Result: &historyservicepb.WorkflowOperationResult_Error{
+					// TODO: improve error reporting
+					Error: &status.Status{Code: code.Code_value[code.Code_INTERNAL.String()]},
+				},
+			}), nil
+	}
+	startResult := &historyservicepb.WorkflowOperationResult{
+		Result: &historyservicepb.WorkflowOperationResult_StartWorkflow{
+			StartWorkflow: startResp,
+		},
+	}
+
+	if updateErr != nil {
+		updateResp := updater.OnError(updateErr)
+		return makeResponse(
+			startResult,
+			&historyservicepb.WorkflowOperationResult{
+				Result: &historyservicepb.WorkflowOperationResult_UpdateWorkflow{
+					UpdateWorkflow: updateResp,
+				},
+			},
+		), nil
+	}
+
+	// TODO: eventually, we'll want to put the MS into the cache as well
+	if err = workflowConsistencyChecker.GetWorkflowCache().Put(
+		shardContext,
+		namespaceID,
+		&commonpb.WorkflowExecution{WorkflowId: req.WorkflowId, RunId: startResp.RunId},
+		workflowCtx,
+		shardContext.GetMetricsHandler(),
+	); err != nil {
+		return nil, err
+	}
```

Comment:
Without this, there's no Update registry on the Matching call back to History.

Reply:
This is worth to be comment in the code, not on PR.